### PR TITLE
Leo Add more specific link validation for GDoc links

### PR DESCRIPTION
**Pull request by Chuehleo**
# Description

## Main changes explained:
- Updated EditLinkModal component to include validation for Google Doc URLs using isValidGoogleDocsUrl.
- Modified validation logic to accept any valid URL for the Media Folder input field.
…

## How to test:
1. check into current branch
2. do `npm install` and `...` to run this PR locally
3. Clear site data/cache
4. log as admin user
5. go to nav bar→ Welcome, XXX→View Profile→Edit(Under the Profile pic)
6. enter various URLs in the Google Doc link to test if only google doc links are allowed
7. verify that any valid URL can be entered in the Media Folder link

## Screenshots or videos of changes:

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/122568562/79f7152b-8259-49e6-9aad-92bc81ba3aaa

### CODE REVIEW

**Review by Sophie-lei — DISMISSED** [review]
Hey, I tested this PR, and everything works as expected. I logged in as an owner role, it can be added Google Doc button with only the Google Doc link. and any URL can be used in the Media Folder link.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/70113822/72a87752-45a4-4dfa-8472-6007654813a5

**Review by ShengjieMao — DISMISSED** [review]
The two functions worked as expected. Tested different input types and both text fields worked great based on criteria

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/56333697/a48a9ade-216e-4aa4-a6fe-ba8a4edcb3aa

**Review by n0vay — DISMISSED** [review]
looks Good to me
image

**Review by Pandani07 — DISMISSED** [review]
I tested this frontend PR with the latest development backend and the feature works as intended. I could only add google-doc links to the google doc links prompt. I could add other links to the media folder prompt and the links were saved.


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/44672833/0b57fa80-a7d9-4399-990f-7080ea4d1235

**Comment by swathyjayaseelan** [comment]
Works as expected.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/13806470/96c3bf59-be95-4eaf-b5f5-8b3aa9544cd7

**Review by prnvjn — DISMISSED** [review]
Hey Leo, tested your PR and the functionality works as described. Tested it using the admin role. 

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/3525231/b866ae42-13f2-48f9-bbfc-41bef72f1d0e

**Review by ClemarNunes — DISMISSED** [review]
I tested this PR and it is working as expected.


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/134466562/f42cdc85-c64c-442b-ba6d-5f1504996de6

**Review by X1aoZhang99 — DISMISSED** [review]
Hi Leo,

I've tested your PR, and all functionalities worked as intended. Please refer to the video.  Nice work! 

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/113070229/b9247a5a-8497-46a7-bd11-a9fc62741a27

**Review by 666saofeng — APPROVED** [review]
Hi, I tested with the admin account and it worked as intended. Great work! Please refer to the video.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/62363348/5632e50d-60b2-4a2f-87f4-cdcbed12cb9d

**Review by peterson337 — DISMISSED** [review]
Chuehleo, I tested your application, and everything worked as expected. Good job!

one comunnity

**Review by yaow62023 — DISMISSED** [review]
I have test your PR, only specify google docs can be added to there. 

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/137429365/cfda269e-a3eb-47ad-9ce8-15f8e8629ccc

File: src/components/UserProfile/UserProfileModal/EditLinkModal.jsx [hunk 1]
```diff
@@ -58,6 +58,18 @@ const EditLinkModal = props => {
   const [isMediaFolderLinkChanged, setIsMediaFolderLinkChanged] = useState(false);
   const [isValidLink, setIsValidLink] = useState(true);
 
+  const isValidGoogleDocsUrl = (url) => {
+    const pattern = /^https:\/\/docs\.google\.com\/document\/d\/[a-zA-Z0-9-_]+/;
+    return pattern.test(url);
+  };
+
+  const isValidMediaUrl = (url) => {
+    // This is a placeholder. You should define what constitutes a valid media URL for your application.
```
Comment (navneeeth) [inline src/components/UserProfile/UserProfileModal/EditLinkModal.jsx:67]: This ChatGPT-generated comment needs to be removed

File: src/components/UserProfile/UserProfileModal/EditLinkModal.jsx [hunk 2]
```diff
@@ -145,37 +157,25 @@ const EditLinkModal = props => {
...
-      handleSubmit();
+    // Validate the Google Doc and Media Folder links
+    const isGoogleDocsValid = googleLink.Link === '' || isValidGoogleDocsUrl(googleLink.Link);
+    const isMediaFolderValid = mediaFolderLink.Link === '' || isValidMediaUrl(mediaFolderLink.Link);
+
+    if (isGoogleDocsValid && isMediaFolderValid) {
+      // If both URLs are valid, proceed with the update logic
+      const linksToUpdate = [googleLink, mediaFolderLink, ...adminLinks];
+      await updateLink(personalLinks, linksToUpdate, mediaFolderLink.Link);
+      handleSubmit(); // If you have additional logic to handle after submitting the form
```
Comment (navneeeth) [inline src/components/UserProfile/UserProfileModal/EditLinkModal.jsx:168]: This ChatGPT-generated comment needs to be removed

File: src/components/UserProfile/UserProfileModal/EditLinkModal.jsx [hunk 3]
```diff
@@ -72,18 +84,18 @@ const EditLinkModal = props => {
...
       setIsChanged(true);
-    } 
+    }
     else {
       setMediaFolderLink({ ...mediaFolderLink, Link: e.target.value.trim() });
       setIsChanged(true);
-      if (!isMediaFolderLinkChanged && !isWarningPopupOpen){ // Fisrt time media folder link is changed
-          setIsMediaFolderLinkChanged(true);
-          setIsWarningPopupOpen(true);
+      if (!isMediaFolderLinkChanged && !isWarningPopupOpen) { // Fisrt time media folder link is changed
```
Comment (navneeeth) [inline src/components/UserProfile/UserProfileModal/EditLinkModal.jsx:96]: Fix the typographical error in the comment: 'Fisrt'

**Review by navneeeth — CHANGES_REQUESTED** [review]
Hello, @Chuehleo! I tested your PR and the functionality works nearly as expected. Please address the changes that I have requested in the Observations and I will approve the PR. Thank you. 

## Observations
- The RegEx to update the Google Doc link works strictly and does not allow regular URLs to be entered as the Google Doc Link.
- The RegEx for the Media Folder allows a good range of permissible URLs in its RegEx design.
- The Google Doc link, however, does not accept 'http' or 'www' as the start of the Google document link which needs to be updated in the RegEx check. Please update the RegEx check to allow links that begin with 'http' and 'www'.
http
www

- The above error message: `Please ensure each link has a unique and not empty, and enter valid URLs.` is not grammatically correct. Please update it as: `Please enter valid URLs for each link.` in the paragraph tag in line 382: 
```javascript
<p className='invalid-help-context' data-testid='invalid-url-warning' >
                  Please ensure each link has a unique and not empty, and enter valid URLs.
                </p>
```
- In line 17, we are importing the following functions:
```javascript
import { isValidGoogleDocsUrl, isValidMediaUrl } from 'utils/checkValidURL';
```
For maintaining best practices, please update your implementation of the isValidGoogleDocsUrl and isValidMediaUrl functions directly in `src/utils/checkValidURL.js` instead of defining them again in the `src/components/UserProfileModal/EditLinkModal.jsx` file. Please also move the following implementation of `isValidUrl` into the `src/utils/checkValidURL.js` file while also including it in the import:
```javascript
const isValidUrl = url => {
    try {
      const pattern = /^(?:https?:\/\/)?[\w.-]+\.[a-zA-Z]{2,}(?:\/\S*)?$/;
      return pattern.test(url);
    } catch (err) {
      console.log(err);
      return false;
    }
  };
```
- The code needs to be proofread as it contains 2 AI-generated comments and a typographical error which need to be removed. I have highlighted them in my review with single-line comments.

### Testing Video

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/44264834/8facf55a-341c-40d8-95c2-cb84742893a2

**Comment by Weiyao-Li** [comment]
Check 1, invalid url:
image

image


Check2, with the correct google doc link:
image

Check3: invalid integer:
image

**Review by Weiyao-Li — COMMENTED** [review]
Reviewed

**Review by Weiyao-Li — DISMISSED** [review]
Looks good, more detailed review in the comment

**Review by Tiantian-C — COMMENTED** [review]
Hi, I have tested this, it works well!
Screenshot 2024-03-12 at 10 36 18 PM
Screenshot 2024-03-12 at 10 36 45 PM

**Comment by rajnada** [comment]
Hi,
I have checked the functionality. It works as mentioned. 


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/25640436/4e0d9970-8e9d-4bad-80fa-2b53d07e4558

**Comment by one-community** [comment]
Thank you all, merging!

**Pull request merged**